### PR TITLE
Get sp app with app-id and tenant domain

### PR DESCRIPTION
**Pull request by sadilchamishka**
### Proposed changes in this pull request

The SaaS apps can be shared across sub organizations by the admin user who owned the SaaS app. The fragment apps created after the organization sharing also owned by the admin of the SaaS app. When retrieving the fragment applications, the app owner's tenant domain can't be used anymore as the fragmented app is assigned against the corresponding sub organization. (In the SP_INBOUND_AUTH table)

Introduced a new method to get the Service provider by the applicationId and tenant domain it resides.

### Related Issues
- https://github.com/wso2/product-is/issues/14335

### CODE REVIEW

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 1]
```diff
@@ -4870,7 +4892,7 @@ public ServiceProvider getApplicationByResourceId(String resourceId,
 
         try {
             int appId = getAppIdUsingResourceId(resourceId, tenantDomain);
-            ServiceProvider application = getApplication(appId);
+            ServiceProvider application = getApplication(appId, tenantDomain);
```
Comment (sadilchamishka) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:4895]: Add a comment why the change is done

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java [hunk 2]
```diff
@@ -55,6 +55,18 @@ ServiceProvider getApplication(String applicationName, String tenantDomain)
      */
     ServiceProvider getApplication(int applicationId) throws IdentityApplicationManagementException;
 
+    /**
+     * @param applicationId
+     * @param tenantDomain
+     * @return
+     * @throws IdentityApplicationManagementException
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java:62]: complete the comment

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java [hunk 3]
```diff
@@ -53,8 +53,21 @@ ServiceProvider getApplication(String applicationName, String tenantDomain)
      * @return
      * @throws IdentityApplicationManagementException
      */
+    @Deprecated
     ServiceProvider getApplication(int applicationId) throws IdentityApplicationManagementException;
 
+    /**
+     * @param applicationId The application id.
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java:60]: missing method description

Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java:65]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4137/commits/9651c36ec3ca6f1e02c745e5db1d728ca80a90aa

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 4]
```diff
@@ -2118,7 +2119,28 @@ public ServiceProvider getApplication(int applicationId) throws IdentityApplicat
...
+            IdentityApplicationManagementUtil.closeConnection(connection);
+        }
+    }
+
+    @Override
+    public ServiceProvider getApplication(int applicationId, String tenantDomain)
+            throws IdentityApplicationManagementException {
+
+        Connection connection = IdentityDatabaseUtil.getDBConnection(false);
+        try {
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:2136]: Let's use try with resource

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java [hunk 5]
```diff
@@ -49,12 +49,30 @@ ServiceProvider getApplication(String applicationName, String tenantDomain)
...
      */
+    @Deprecated
     ServiceProvider getApplication(int applicationId) throws IdentityApplicationManagementException;
 
+    /**
+     * Get service provider resides in given tenant domain.
+     *
+     * @param applicationId The application id.
+     * @param tenantDomain  The tenant domain of the application resides.
+     * @return
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java:67]:
```suggestion
     * @return Service provider.
```

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java [hunk 6]
```diff
@@ -49,12 +49,30 @@ ServiceProvider getApplication(String applicationName, String tenantDomain)
             throws IdentityApplicationManagementException;
 
     /**
+     * Get service provider when the application resides in the same tenant of the request initiated.
+     *
      * @param applicationId
      * @return
      * @throws IdentityApplicationManagementException
+     * @deprecated please use #getApplication(applicationId, tenantDomain)
```
Comment (AnuradhaSK) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java:57]:
```suggestion
     * @deprecated with getApplication(applicationId, tenantDomain) method.
```

Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4137/commits/0095387c7b31cfe863481084d216693e001f50fa

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java (hunk 5 above)
Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4137/commits/0095387c7b31cfe863481084d216693e001f50fa

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java (hunk 4 above)
Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4137/commits/0095387c7b31cfe863481084d216693e001f50fa

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 7]
```diff
@@ -2118,7 +2119,28 @@ public ServiceProvider getApplication(int applicationId) throws IdentityApplicat
...
+
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4137/commits/0095387c7b31cfe863481084d216693e001f50fa

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java [hunk 8]
```diff
@@ -49,12 +49,30 @@ ServiceProvider getApplication(String applicationName, String tenantDomain)
...
+    @Deprecated
     ServiceProvider getApplication(int applicationId) throws IdentityApplicationManagementException;
 
+    /**
+     * Get service provider resides in given tenant domain.
+     *
+     * @param applicationId The application id.
+     * @param tenantDomain  The tenant domain of the application resides.
+     * @return Service provider.
+     * @throws IdentityApplicationManagementException
```
Comment (dewniMW) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java:68]: explain in which instances this exception will occur

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java [hunk 9]
```diff
@@ -49,12 +49,30 @@ ServiceProvider getApplication(String applicationName, String tenantDomain)
...
+     * @deprecated with getApplication(applicationId, tenantDomain) method.
      */
+    @Deprecated
     ServiceProvider getApplication(int applicationId) throws IdentityApplicationManagementException;
 
+    /**
+     * Get service provider resides in given tenant domain.
+     *
+     * @param applicationId The application id.
+     * @param tenantDomain  The tenant domain of the application resides.
```
[the remark this answers is not in the document]
Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/ApplicationDAO.java]: Fixed by - https://github.com/wso2/carbon-identity-framework/pull/4137/commits/e7279da49affb559f419f24f48d3eed56ceaa84e

**Review by dewniMW — APPROVED** [review]

File: components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java [hunk 10]
```diff
@@ -2118,7 +2119,26 @@ public ServiceProvider getApplication(int applicationId) throws IdentityApplicat
...
+            IdentityApplicationManagementUtil.closeConnection(connection);
+        }
+    }
+
+    @Override
+    public ServiceProvider getApplication(int applicationId, String tenantDomain)
+            throws IdentityApplicationManagementException {
+
+        try (Connection connection = IdentityDatabaseUtil.getDBConnection(false)) {
+            // Load basic application data
```
Comment (IsankaSR) [inline components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:2136]: Missing full stop at the end

Comment (sadilchamishka) [reply components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/dao/impl/ApplicationDAOImpl.java:2136]: The work of this PR for Introducing a new method and deprecating the previous method was revert back by another PR https://github.com/wso2/carbon-identity-framework/pull/4157/files#diff-0bd1a832f4495aed005c591463b854143fff7695f721b85b5d52f6fecdfb2f63L2136
